From 40e236d9f225288d99577223f35aed67173fbc01 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Fri, 4 Jul 2014 13:28:45 +1000
Subject: [PATCH] Further Seed Customisation

Allow server admins that really want to to customise the seeds used in world generation even further.

diff --git a/src/main/java/net/minecraft/server/ChunkGenerator.java b/src/main/java/net/minecraft/server/ChunkGenerator.java
index cea5166a..4b3a823f 100644
--- a/src/main/java/net/minecraft/server/ChunkGenerator.java
+++ b/src/main/java/net/minecraft/server/ChunkGenerator.java
@@ -196,4 +196,10 @@ public abstract class ChunkGenerator<C extends GeneratorSettingsDefault> {
     public int c(int i, int j, HeightMap.Type heightmap_type) {
         return this.getBaseHeight(i, j, heightmap_type) - 1;
     }
+
+    // Spigot start
+    public World getWorld() {
+        return this.a.getMinecraftWorld();
+    }
+    // Spigot end
 }
diff --git a/src/main/java/net/minecraft/server/EntitySlime.java b/src/main/java/net/minecraft/server/EntitySlime.java
index 54824842..3140a49e 100644
--- a/src/main/java/net/minecraft/server/EntitySlime.java
+++ b/src/main/java/net/minecraft/server/EntitySlime.java
@@ -276,7 +276,7 @@ public class EntitySlime extends EntityInsentient implements IMonster {
                 }
 
                 ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(blockposition);
-                boolean flag = SeededRandom.a(chunkcoordintpair.x, chunkcoordintpair.z, generatoraccess.getSeed(), 987234911L).nextInt(10) == 0;
+                boolean flag = SeededRandom.a(chunkcoordintpair.x, chunkcoordintpair.z, generatoraccess.getSeed(), world.spigotConfig.slimeSeed).nextInt(10) == 0; // Spigot
 
                 if (this.random.nextInt(10) == 0 && flag && this.locY < 40.0D) {
                     return super.a(generatoraccess, enummobspawn);
diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureDesertPyramid.java b/src/main/java/net/minecraft/server/WorldGenFeatureDesertPyramid.java
index 4f1d57b8..c5448463 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureDesertPyramid.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureDesertPyramid.java
@@ -25,8 +25,10 @@ public class WorldGenFeatureDesertPyramid extends WorldGenFeatureRandomScattered
     }
 
     @Override
-    protected int getSeed() {
-        return 14357617;
+    // Spigot start
+    protected int getSeed(World world) {
+        return world.spigotConfig.desertSeed;
+        // Spigot end
     }
 
     public static class a extends StructureStart {
diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureIgloo.java b/src/main/java/net/minecraft/server/WorldGenFeatureIgloo.java
index 0f6ce152..6f8a9794 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureIgloo.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureIgloo.java
@@ -25,8 +25,10 @@ public class WorldGenFeatureIgloo extends WorldGenFeatureRandomScattered<WorldGe
     }
 
     @Override
-    protected int getSeed() {
-        return 14357618;
+    // Spigot start
+    protected int getSeed(World world) {
+        return world.spigotConfig.iglooSeed;
+        // Spigot end
     }
 
     public static class a extends StructureStart {
diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureJunglePyramid.java b/src/main/java/net/minecraft/server/WorldGenFeatureJunglePyramid.java
index 35458425..ca6b1ee8 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureJunglePyramid.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureJunglePyramid.java
@@ -25,8 +25,10 @@ public class WorldGenFeatureJunglePyramid extends WorldGenFeatureRandomScattered
     }
 
     @Override
-    protected int getSeed() {
-        return 14357619;
+    // Spigot start
+    protected int getSeed(World world) {
+        return world.spigotConfig.jungleSeed;
+        // Spigot end
     }
 
     public static class a extends StructureStart {
diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureOceanRuin.java b/src/main/java/net/minecraft/server/WorldGenFeatureOceanRuin.java
index 5f847e68..7d553ee4 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureOceanRuin.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureOceanRuin.java
@@ -39,8 +39,10 @@ public class WorldGenFeatureOceanRuin extends WorldGenFeatureRandomScattered<Wor
     }
 
     @Override
-    protected int getSeed() {
-        return 14357621;
+    // Spigot start
+    protected int getSeed(World world) {
+        return world.spigotConfig.oceanSeed;
+        // Spigot end
     }
 
     public static enum Temperature {
diff --git a/src/main/java/net/minecraft/server/WorldGenFeaturePillagerOutpost.java b/src/main/java/net/minecraft/server/WorldGenFeaturePillagerOutpost.java
index 9a0dd37e..69b3f404 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeaturePillagerOutpost.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeaturePillagerOutpost.java
@@ -67,8 +67,10 @@ public class WorldGenFeaturePillagerOutpost extends WorldGenFeatureRandomScatter
     }
 
     @Override
-    protected int getSeed() {
-        return 165745296;
+    // Spigot start
+    protected int getSeed(World world) {
+        return world.spigotConfig.outpostSeed;
+        // Spigot end
     }
 
     public static class a extends StructureAbstract {
diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureRandomScattered.java b/src/main/java/net/minecraft/server/WorldGenFeatureRandomScattered.java
index ae797325..49ae612b 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureRandomScattered.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureRandomScattered.java
@@ -21,7 +21,7 @@ public abstract class WorldGenFeatureRandomScattered<C extends WorldGenFeatureCo
         int k2 = i2 / i1;
         int l2 = j2 / i1;
 
-        ((SeededRandom) random).a(chunkgenerator.getSeed(), k2, l2, this.getSeed());
+        ((SeededRandom) random).a(chunkgenerator.getSeed(), k2, l2, this.getSeed(chunkgenerator.getWorld())); // Spigot
         k2 *= i1;
         l2 *= i1;
         k2 += random.nextInt(i1 - j1);
@@ -52,5 +52,5 @@ public abstract class WorldGenFeatureRandomScattered<C extends WorldGenFeatureCo
         return chunkgenerator.getSettings().i();
     }
 
-    protected abstract int getSeed();
+    protected abstract int getSeed(World world); // Spigot
 }
diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureShipwreck.java b/src/main/java/net/minecraft/server/WorldGenFeatureShipwreck.java
index e89dbbcb..061eb256 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureShipwreck.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureShipwreck.java
@@ -25,8 +25,10 @@ public class WorldGenFeatureShipwreck extends WorldGenFeatureRandomScattered<Wor
     }
 
     @Override
-    protected int getSeed() {
-        return 165745295;
+    // Spigot start
+    protected int getSeed(World world) {
+        return world.spigotConfig.shipwreckSeed;
+        // Spigot end
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureSwampHut.java b/src/main/java/net/minecraft/server/WorldGenFeatureSwampHut.java
index a68795c3..e8750d82 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureSwampHut.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureSwampHut.java
@@ -30,8 +30,10 @@ public class WorldGenFeatureSwampHut extends WorldGenFeatureRandomScattered<Worl
     }
 
     @Override
-    protected int getSeed() {
-        return 14357620;
+    // Spigot start
+    protected int getSeed(World world) {
+        return world.spigotConfig.swampSeed;
+        // Spigot end
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/WorldGenMonument.java b/src/main/java/net/minecraft/server/WorldGenMonument.java
index 2a072e07..053004b8 100644
--- a/src/main/java/net/minecraft/server/WorldGenMonument.java
+++ b/src/main/java/net/minecraft/server/WorldGenMonument.java
@@ -27,7 +27,7 @@ public class WorldGenMonument extends StructureGenerator<WorldGenFeatureEmptyCon
         int k2 = i2 / i1;
         int l2 = j2 / i1;
 
-        ((SeededRandom) random).a(chunkgenerator.getSeed(), k2, l2, 10387313);
+        ((SeededRandom) random).a(chunkgenerator.getSeed(), k2, l2, chunkgenerator.getWorld().spigotConfig.monumentSeed); // Spigot
         k2 *= i1;
         l2 *= i1;
         k2 += (random.nextInt(i1 - j1) + random.nextInt(i1 - j1)) / 2;
diff --git a/src/main/java/net/minecraft/server/WorldGenVillage.java b/src/main/java/net/minecraft/server/WorldGenVillage.java
index ead45107..b155a73a 100644
--- a/src/main/java/net/minecraft/server/WorldGenVillage.java
+++ b/src/main/java/net/minecraft/server/WorldGenVillage.java
@@ -21,7 +21,7 @@ public class WorldGenVillage extends StructureGenerator<WorldGenFeatureVillageCo
         int k2 = i2 / i1;
         int l2 = j2 / i1;
 
-        ((SeededRandom) random).a(chunkgenerator.getSeed(), k2, l2, 10387312);
+        ((SeededRandom) random).a(chunkgenerator.getSeed(), k2, l2, chunkgenerator.getWorld().spigotConfig.villageSeed); // Spigot
         k2 *= i1;
         l2 *= i1;
         k2 += random.nextInt(i1 - j1);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index c9380a3f..b4fcf25d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -148,7 +148,7 @@ public class CraftChunk implements Chunk {
     @Override
     public boolean isSlimeChunk() {
         // 987234911L is deterimined in EntitySlime when seeing if a slime can spawn in a chunk
-        return SeededRandom.a(getX(), getZ(), getWorld().getSeed(), 987234911L).nextInt(10) == 0;
+        return SeededRandom.a(getX(), getZ(), getWorld().getSeed(), worldServer.spigotConfig.slimeSeed).nextInt(10) == 0;
     }
 
     public boolean unload(boolean save) {
diff --git a/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java b/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java
index c9d3f5f4..b474d0bb 100644
--- a/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java
+++ b/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java
@@ -164,4 +164,10 @@ public class CustomChunkGenerator extends InternalChunkGenerator<GeneratorSettin
     public int getGenerationDepth() {
         return world.getHeight();
     }
+
+    // Spigot start
+    public WorldServer getWorld() {
+        return world;
+    }
+    // Spigot end
 }
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index ee6c4d56..462d0cfe 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -230,4 +230,29 @@ public class SpigotWorldConfig
     {
         witherSpawnSoundRadius = getInt( "wither-spawn-sound-radius", 0 );
     }
+
+    public int villageSeed;
+    public int desertSeed;
+    public int iglooSeed;
+    public int jungleSeed;
+    public int swampSeed;
+    public int monumentSeed;
+    public int oceanSeed;
+    public int outpostSeed;
+    public int shipwreckSeed;
+    public int slimeSeed;
+    private void initWorldGenSeeds()
+    {
+        villageSeed = getInt( "seed-village", 10387312 );
+        desertSeed = getInt( "seed-desert", 14357617 );
+        iglooSeed = getInt( "seed-igloo", 14357618 );
+        jungleSeed = getInt( "seed-jungle", 14357619 );
+        swampSeed = getInt( "seed-swamp", 14357620 );
+        monumentSeed = getInt( "seed-monument", 10387313 );
+        shipwreckSeed = getInt( "seed-shipwreck", 165745295 );
+        oceanSeed = getInt( "seed-ocean", 14357621 );
+        outpostSeed = getInt( "seed-outpost", 165745296 );
+        slimeSeed = getInt( "seed-slime", 987234911 );
+        log( "Custom Map Seeds:  Village: " + villageSeed + " Desert: " + desertSeed + " Igloo: " + iglooSeed + " Jungle: " + jungleSeed + " Swamp: " + swampSeed + " Monument: " + monumentSeed + "Ocean: " + oceanSeed + " Shipwreck: " + shipwreckSeed + " Slime: " + slimeSeed );
+    }
 }
-- 
2.20.1

